webanno-api-annotation/src/main/java/de/tudarmstadt/ukp/clarin/webanno/api/annotation/rendering/PreRenderer.java

```diff
@@ -48,13 +47,14 @@ public void render(VDocument aResponse, AnnotatorState aState, JCas aJCas,
         // Render visible (custom) layers
         for (AnnotationLayer layer : aLayers) {
             List<AnnotationFeature> features = annotationService.listAnnotationFeature(layer);
-            List<AnnotationFeature> invisibleFeatures = new ArrayList<>();
-            for (AnnotationFeature feature : features) {
-                if (!feature.isVisible()) {
-                    invisibleFeatures.add(feature);
-                }
-            }
-            features.removeAll(invisibleFeatures);
+            
+            // FIXME: is this really necessary here??
+//            for (AnnotationFeature feature : features) {
+//                if (!feature.isVisible()) {
+//                    invisibleFeatures.add(feature);
+//                }
+//            }
+//            features.removeAll(invisibleFeatures);
```

Comment:
Why not?

Reply:
because for the `renderer.render(aJCas, features, aResponse, aState);` only visible layers are passed, but in  `Renderer. getFeatures(TypeAdapter aAdapter, AnnotationFS aFs, List<AnnotationFeature> aFeatures)` method visibilty is checked anyway:  `if (!feature.isEnabled() || !feature.isVisible()
                    || !MultiValueMode.NONE.equals(feature.getMultiValueMode())) {
                continue;
            }` 
If I'm not missing something, it feels kind of duplicated.

Also, I thought about the hoverFeatures to be independent of visibility,  if we leave it uncommented, we'd need to pass another parameter to the `render` method, like `render(aJCas, visibleFeatures, invisibleFeatures, aResponse, aState)` in order for the `getHoverFeatures` to access all layers (not only the visible ones).

What do you think? Am I missing something? I tested different anntation types in different settings (visibilty on/off,...) and the visualization appears the same, but I might be missing a special case.

Reply:
Better comment this back in and open another issue. Removing this filtering would presently indirectly affect `Renderer.renderRequiredFeatureErrors()` Probably there should not be a `features` parameter at all in `Renderer.render()`. However, looking a the code, removing this parameter will entail further changes and fixes. Better handle it separately.

Reply:
alright, uncommented it (add4fe13515987be7891e60c7b0b1cd87b80d920)

Reply:
... and made it slightly more performant